File: 853-most-profit-assigning-work/most-profit-assigning-work.cpp
class Solution {
public:
    static bool cmp(pair<int ,int>&a , pair<int , int>&b){
        if(a.first == b.first){
            return a.second > b.second;
        }
        return a.first < b.first;
    }
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        vector<pair<int , int>>arr;
        int n = profit.size();
        for(int i = 0 ; i < n; i++){
            arr.push_back({ difficulty[i] , profit[i]});
        }

        sort(arr.begin() , arr.end() , cmp);
        sort(worker.begin() , worker.end());

        int m = worker.size();

        int i = 0;
        int j = 0;
        int sum = 0;
        int maxi = 0;
        while(i < m){
            while(j < n && worker[i] >= arr[j].first){
                maxi = max(maxi , arr[j].second);
                j++;
            }
            sum += maxi;
            i++;

            
        }
        return sum;
        
    }
};

File: 853-most-profit-assigning-work/most-profit-assigning-work.java
class Solution {
    public int maxProfitAssignment(int[] difficulty, int[] profit, int[] worker) {
        int n = profit.length;
        int[][] arr = new int [n][2];

        for(int i = 0 ; i < n ; i++){
            arr[i][0] = difficulty[i];
            arr[i][1] = profit[i];
        }

        Arrays.sort(arr , (a , b) ->{
            if(a[0] == b[0]){
                return b[1] - a[1];
            }
            return a[0] - b[0];
        });

        Arrays.sort(worker);

        int lastmax = 0;
        int sum = 0;
        int i = 0 , j = 0;
        while( j< worker.length ){
            if(i < n && arr[i][0] <= worker[j]){
                if(lastmax < arr[i][1]){
                    lastmax = arr[i][1];

                }
                i++;
            }
            else{
                sum += lastmax;
                j++;
            }
        }
        return sum;


        
    }
}